### PR TITLE
remove error() from Expression

### PR DESCRIPTION
**Pull request by WalterBright**
This is a continuation of:

1. simplifying the compile AST's, mainly to combine the two AST trees into one
2. unifying the error message handling into ErrorSink's by getting rid of the rather purposeless method of having error() functions as members of the AST

The changes here are trivial, just a tediously large quantity of them.

### CODE REVIEW

File: compiler/src/dmd/hdrgen.d [hunk 1]
```diff
@@ -92,9 +92,9 @@ extern (C++) void genhdrfile(Module m, ref OutBuffer buf)
  * Turn a Statement into a string suitable for printf.
  * Leaks memory.
  * Params:
- *	s = Statement to convert
+ *      s = Statement to convert
  * Returns:
- *	0-terminated string
+ *      0-terminated string
```
Comment (thewilsonator) [inline compiler/src/dmd/hdrgen.d:97]: unrelated

Comment (WalterBright) [reply compiler/src/dmd/hdrgen.d:97]: That's just a tab a previous PR forgot.

File: compiler/src/dmd/dsymbolsem.d [hunk 2]
```diff
@@ -1737,7 +1737,7 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                     const name = (cast(const(char)[])se.peekData()).xarraydup;
                     uint cnt = setMangleOverride(s, name);
                     if (cnt > 1)
-                        pd.error("can only apply to a single declaration");
```
Comment (dkorpel) [inline compiler/src/dmd/dsymbolsem.d:1740]: This is not an expression but a declaration, and declaration.error also prints the pragma declaration here

Comment (WalterBright) [reply compiler/src/dmd/dsymbolsem.d]: fixed

**Comment by WalterBright** [comment]
@benjones I'd like to keep the functions in ErrorSink at a minimum.

**Comment by WalterBright** [comment]
@dkorpel ready to rock

File: compiler/test/fail_compilation/fail235.d [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /*
 TEST_OUTPUT:
 ---
-fail_compilation/fail235.d(12): Error: expression `typeid(char)` is not a valid template value argument
+fail_compilation/fail235.d(12): Error: template instance `Tuple!(typeid(char))` expression `typeid(char)` is not a valid template value argument
```
Comment (dkorpel) [inline compiler/test/fail_compilation/fail235.d:4]: Why these test changes?

**Review by dkorpel — APPROVED** [review]

**Comment by maxhaton** [comment]
> I'd like to keep the functions in ErrorSink at a minimum.

Why? Its one tiny function versus making the whole tree uglier + more prone to bugs.

By doing it this way you've made the code much more brittle. No one should ever have to access .loc directly, it should be just for downstream - there is always an unknown-unknown benefit waiting if you have a sensible API that lets people do transformative work underneath you, but with this approach you're basically one step away from an inline printf call.

**Comment by benjones** [comment]
> > I'd like to keep the functions in ErrorSink at a minimum.
> 
> Why? Its one tiny function versus making the whole tree uglier + more prone to bugs.
> 
> By doing it this way you've made the code much more brittle. No one should ever have to access .loc directly, it should be just for downstream - there is always an unknown-unknown benefit waiting if you have a sensible API that lets people do transformative work underneath you, but with this approach you're basically one step away from an inline printf call.

Half-finished attempt at actually doing it here: https://github.com/dlang/dmd/pull/15653

File: compiler/test/fail_compilation/fail235.d (hunk 3 above)
Comment (WalterBright) [reply compiler/test/fail_compilation/fail235.d:4]: Variations in the way the different error functions worked. The current system of a patchwork of slightly different functions is slowly being replaced with one interface, ErrorSink. This PR is progress in that direction.

**Comment by WalterBright** [comment]
> Its one tiny function versus making the whole tree uglier + more prone to bugs

The current system is a patchwork of arbitrary functions for reporting errors. This PR is a step towards harmonizing it.

**Comment by WalterBright** [comment]
> Its one tiny function versus making the whole tree uglier + more prone to bugs.

I agree with the goal of making it less prone to bugs, I don't see how simplifying the error reporting system leads to that.

**Comment by tgehr** [comment]
I don't really see the benefit of removing simple helper functions. The main thing this does is it breaks all PRs. Also, if at any point in the future you want to get rid of the global error sink, you will break all PRs again.
